### PR TITLE
Feature/deserde

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: Cargo.toml
Comment (VictorKoenders) [inline Cargo.toml]: Should we rename this to `-alpha.0` before merging into trunk? Semver pre-release versions are alphabetically ordered  which means `-dev` is after `-alpha`.

Alternatively we can call it `-aaaa` to make sure it's always older

File: Cargo.toml
Comment (VictorKoenders) [inline Cargo.toml]: Should be `# Used for tests`

File: derive/src/error.rs
Comment (VictorKoenders) [inline derive/src/error.rs]: TODO (me): we can use the generator API for this

File: derive/src/generate/generate_fn.rs
Comment (VictorKoenders) [inline derive/src/generate/generate_fn.rs:5]: Nitpick: newline between `use ...` and `struct ...`

File: derive/src/generate/generator.rs
Comment (VictorKoenders) [inline derive/src/generate/generator.rs:14]: TODO (me): I think these functions need to be documented, especially because I'm considering making this a seperate crate, but I won't block this PR on it

File: derive/src/generate/stream_builder.rs
Comment (VictorKoenders) [inline derive/src/generate/stream_builder.rs]: TODO (me): this should probably return a `Result`

File: docs/spec.md
Comment (VictorKoenders) [inline docs/spec.md:1]: @ZoeyR could you review this document to see if I missed something?

File: docs/spec.md
Comment (VictorKoenders) [inline docs/spec.md:57]: This should explicitly mention that it's serialized as a `u32`

File: docs/spec.md
Comment (VictorKoenders) [inline docs/spec.md]: The examples in this document should match the new `Config` api.

Comment (VictorKoenders) [reply docs/spec.md]: Also I think we can include this document in rustdoc with `#[doc = include_str!("../docs/spec.md")]` somewhere in the source code. This will help with testing the examples and it'll show up on docs.rs

File: docs/spec.md
Comment (VictorKoenders) [inline docs/spec.md:97]: This is dependent on a config flag now

File: docs/spec.md
Comment (VictorKoenders) [inline docs/spec.md]: Same here

File: src/de/decoder.rs
Comment (VictorKoenders) [inline src/de/decoder.rs:24]: We were considering renaming this to `DecoderState`, as well as adding the number of bytes written which we'll need for the `byte limit` config setting

File: src/de/impls.rs
Comment (VictorKoenders) [inline src/de/impls.rs:379]: Could `'a` and `'de` be combined into a single lifetime here?

File: src/de/read.rs
Comment (VictorKoenders) [inline src/de/read.rs]: This function is only used for `BorrowReader::take_bytes` and nowhere else, what was the reason we split this again?

File: src/enc/encoder.rs
Comment (VictorKoenders) [inline src/enc/encoder.rs]: This documentation should be updated

File: src/enc/mod.rs
Comment (VictorKoenders) [inline src/enc/mod.rs]: This documentation should be updated. This trait is implemented for every type that needs to be encodable.

File: src/enc/mod.rs
Comment (VictorKoenders) [inline src/enc/mod.rs]: This `sealed` mod should probably be moved to a central utils file, we use Sealed in more places

File: src/features/impl_std.rs
Comment (VictorKoenders) [inline src/features/impl_std.rs]: I'm thinking we should remove all `encode_into_xxx_with_config` and `decode_from_xxx_with_config` and always make `Config` a mandatory selection by the user.

We'll want to update `Config` in the future for more effective encoding/decoding options, if we provide people with a default in the form of `encode_into_write` which uses `Configuration::standard()` on bincode 2.0 launch, we can never change the default to anything else than `::standard()`. Making the user specify the `Configuration` explicitly will make them more aware that there are new config options available in the future.

Comment (VictorKoenders) [reply src/features/impl_std.rs]: In addition it'll halve the amount of encode/decode options we have, which will be important because e.g. the `serde` feature will probably double the amount of encode/decode functions (one for `&mut [u8]`, one for `std::io::Write`, etc)

File: src/features/impl_std.rs
Comment (VictorKoenders) [inline src/features/impl_std.rs]: TODO(me): Copy paste error, should be `RwLock`

File: src/features/impl_std.rs
Comment (VictorKoenders) [inline src/features/impl_std.rs]: I introduced the `EncodeError::Other` variant here because I didn't know how many variants we'd have of weird corner cases. I think we can get rid of `Other` for now and make it specifically an `InvalidPathCharacters` error

File: src/lib.rs
Comment (VictorKoenders) [inline src/lib.rs]: This is now renamed to `Configuration::standard()`. I'm surprised the CI did not catch this

File: src/features/impl_std.rs
Comment (ghost) [reply src/features/impl_std.rs]: I think we still want to keep an "Other" variant. 3rd parties may implement `Encode` and `Decode` for types that have their own correctness rules.

File: src/features/impl_std.rs
Comment (ghost) [reply src/features/impl_std.rs]: Wouldn't the serde solution be to just implement `Encode` and `Decode` on any type that implements `Serialize` and `Deserialize`? Then we wouldn't need an extra entry point.

File: src/de/impls.rs
Comment (ghost) [reply src/de/impls.rs:379]: That would actually make the implementation more specific than what we want. If the lifetime was `&'de [u8]` then you could only decode a slice that lived as long as the backing storage. You would not be able to decode a slice that lived for a shorter time, which will often be the case.

File: src/de/read.rs
Comment (ghost) [reply src/de/read.rs]: I'm not sure, but it might be needed for the `peek_read` optimization that got added in trunk. I'll evaluate.

File: docs/spec.md
Comment (VictorKoenders) [reply docs/spec.md]: Currently we do doctests in a dedicated `specs` module, however because the doctests require both `alloc` and `derive`, we disable the `specs` module when those features are not enabled.

It would be nice if we can ignore the doctests if those features are disabled but still compile the documentation for the user, however I don't know how to do that

File: src/features/impl_std.rs
Comment (VictorKoenders) [reply src/features/impl_std.rs]: I'm not sure if that's possible. If it is, then there won't be that issue. However I still think we don't need dedicated `_with_config` methods

File: src/features/impl_std.rs
Comment (VictorKoenders) [reply src/features/impl_std.rs]: Added the new variant but kept `Other`

**Review by VictorKoenders — APPROVED** [review]

**Pull request merged**
